modules/basicfuncs/numeric-funcs.c

```diff
@@ -87,19 +106,24 @@ parse_integer_or_float(const char *str, Number *number)
 void
 format_number(GString *result, Number n)
 {
-  if (n.value_type == Integer)
+  if (n.precision == 0)
     {
-      format_int64_padded(result, 0, ' ', 10, number_as_int(n));
+      if (n.value_type == Integer)
+        {
+          format_int64_padded(result, 0, ' ', 10, number_as_int(n));
+          return;
+        }
+
+      /*
+       * This special case is required because:
+       * round(0.5)          #=> 1
+       * printf("%.0f", 0.5) #=> 0
```

Comment:
I am still investigating the reason of this, and hope to simplify this code when I fully understand what is going on here.

Reply:
The [POSIX specification of the 'f' format string](https://pubs.opengroup.org/onlinepubs/009695399/functions/fprintf.html) says (emphasis by myself):

> f, F
>     The double argument shall be converted to decimal notation in the style "[-]ddd.ddd", where the number of digits after the radix character is equal to the precision specification. If the precision is missing, it shall be taken as 6; if the precision is explicitly zero and no '#' flag is present, no radix character shall appear. If a radix character appears, at least one digit appears before it. **The low-order digit shall be rounded in an implementation-defined manner.**

So le'ts use `round(3)` when possible, and rely on the implementation-defined rounding otherwise.

Reply:
Do you have an opinion about this?

1) Is this okay as it is currently?
2) Should we always round using %f (that is the test suite needs to be updated because 0.5 is rounded to 0)?
3) Should we always use round(3) (mutiply the number by 10^precision, do the rounding, divide by 10^precision and display this result with %f and precision digits)?

More and more I think 3 is the best way to go because of the consistency of the roundings, but 1 is ways more simple and maybe acceptable for a logging software. 2 is hybrid and inconsistent, so maybe it's the worst possibility.

What do you think?

Reply:
I can agree on the 3rd option, as precision cannot be a conserve while it is limited to 20 digits from the 300 something `gdouble` has, in addition to the reasons you wrote.

Reply:
I left a note for the wip patch you linked.